src/core/ucc_coll.c

```diff
@@ -222,28 +258,45 @@ UCC_CORE_PROFILE_FUNC(ucc_status_t, ucc_collective_post, (request),
                       ucc_coll_req_h request)
 {
     ucc_coll_task_t *task = ucc_derived_of(request, ucc_coll_task_t);
-
+    ucc_status_t status;
     ucc_debug("coll_post: req %p, seq_num %u", task, task->seq_num);
 
     COLL_POST_STATUS_CHECK(task);
     if (UCC_COLL_TIMEOUT_REQUIRED(task)) {
         task->start_time = ucc_get_time();
     }
 
+    if (task->flags & UCC_COLL_TASK_FLAG_EXECUTOR) {
+        status = ucc_ee_executor_start(task->executor, NULL);
+        if (ucc_unlikely(status != UCC_OK)) {
+            ucc_error("failed to start executor: %s",
+                      ucc_status_string(status));
+        }
+    }
     return task->post(task);
 }
 
 UCC_CORE_PROFILE_FUNC(ucc_status_t, ucc_collective_finalize, (request),
                       ucc_coll_req_h request)
 {
     ucc_coll_task_t *task = ucc_derived_of(request, ucc_coll_task_t);
+    ucc_status_t task_st, exec_st;
 
     ucc_debug("coll_finalize: req %p, seq_num %u", task, task->seq_num);
-    if (task->super.status == UCC_INPROGRESS) {
+    if (ucc_unlikely(task->super.status == UCC_INPROGRESS)) {
         ucc_error("attempt to finalize task with status UCC_INPROGRESS");
         return UCC_ERR_INVALID_PARAM;
     }
-    return task->finalize(task);
+
+    task_st = task->finalize(task);
```

Comment:
shouldn't you finalize executor first and only then call task->finalize ? otherwise you might have garbage in task->executor.